File: app/src/lib/__tests__/sshSessions.test.ts
import {
  attachSshSink,
  closeSshHost,
  closeSshTab,
  openSshTab,
  renameSshTab,
  resizeSshHost,
  selectSshTab,
  sshHostSized,
  sshHostStarted,
  sshHostState,
  subscribeSshHost,
} from '../sshSessions';
import { remotlySsh } from '../ssh';
import terminalStoreModule from '../../specs/NativeRemotlyTerminalStore';
import { decodeBase64 } from '../base64';

const terminalStore = terminalStoreModule as unknown as {
  feed: jest.Mock;
  has: jest.Mock;
};

/** Lets the queued writes settle; the drain is asynchronous. */
async function flushMicrotasks(): Promise<void> {
  for (let i = 0; i < 20; i += 1) await Promise.resolve();
}

// Untyped inside the factory on purpose: jest hoists it above the imports and
// rejects any reference out of its own scope, which includes a type name.
// The typed view of it is `ssh` below.
// The store writes into a session's own terminal when no view is attached.
// The fake records what it was handed so the tests can assert nothing is lost.
jest.mock('../../specs/NativeRemotlyTerminalStore', () => {
  const written = new Map();
  return {
    __esModule: true,
    default: {
      feed: jest.fn(async (sessionId, data) => {
        const prev = written.get(sessionId) ?? [];
        prev.push(data);
        written.set(sessionId, prev);
        return true;
      }),
      has: jest.fn(async () => true),
    },
    __written: written,
  };
});

jest.mock('../ssh', () => {
  // Untyped inside the factory on purpose: jest hoists it above the imports
  // and rejects any name from outside its own scope, which a type annotation
  // in a generic argument also trips. The typed view of it is `ssh` below.
  const handlers = new Map();
  return {
    __esModule: true,
    stageMessage: () => null,
    remotlySsh: {
      connect: jest.fn().mockResolvedValue(undefined),
      write: jest.fn().mockResolvedValue(undefined),
      resize: jest.fn().mockResolvedValue(undefined),
      hostKey: jest.fn().mockResolvedValue(undefined),
      close: jest.fn().mockResolvedValue(undefined),
      closeHost: jest.fn().mockResolvedValue(undefined),
      onState: jest.fn(() => () => undefined),
      onData: jest.fn((h: unknown, s: unknown, handler: unknown) => {
        handlers.set(`${h}:${s}`, handler);
        return () => handlers.delete(`${h}:${s}`);
      }),
      __emit: (h: unknown, s: unknown, payload: unknown, fastPath = false) => {
        handlers.get(`${h}:${s}`)?.(payload, fastPath);
      },
    },
  };
});

type MockSsh = typeof remotlySsh & {
  __emit: (
    hostId: string,
    sessionId: string,
    bytes: Uint8Array,
    fastPath?: boolean,
  ) => void;
};
const ssh = remotlySsh as MockSsh;

let host = 0;
function freshHost(): string {
  host += 1;
  return `h${host}`;
}

beforeEach(() => {
  jest.clearAllMocks();
});

describe('session lifetime', () => {
  // The regression this store exists for: a screen unmounting on back must not
  // take the shell with it.
  it('keeps sessions after the renderer detaches', () => {
    const id = freshHost();
    const detach = attachSshSink(id, () => undefined);
    openSshTab(id);
    expect(sshHostState(id).tabs).toHaveLength(1);

    detach();

    expect(sshHostState(id).tabs).toHaveLength(1);
    expect(ssh.close).not.toHaveBeenCalled();
  });

  it('does not reopen a tab for a host that already has one', () => {
    const id = freshHost();
    openSshTab(id);
    expect(sshHostStarted(id)).toBe(true);
    expect(sshHostState(id).tabs).toHaveLength(1);
  });

  it('closes exactly the tab that was asked for', () => {
    const id = freshHost();
    openSshTab(id);
    openSshTab(id);
    const [first, second] = sshHostState(id).tabs;

    closeSshTab(id, first.sessionId);

    expect(ssh.close).toHaveBeenCalledWith(id, first.sessionId);
    expect(sshHostState(id).tabs.map(t => t.sessionId)).toEqual([
      second.sessionId,
    ]);
  });

  it('closes every tab when the host disconnects', () => {
    const id = freshHost();
    openSshTab(id);
    openSshTab(id);

    closeSshHost(id);

    expect(ssh.close).toHaveBeenCalledTimes(2);
    expect(sshHostState(id).tabs).toHaveLength(0);
  });
});

describe('output routing', () => {
  it('delivers the active tab output to the attached renderer', () => {
    const id = freshHost();
    const seen: number[] = [];
    attachSshSink(id, b => seen.push(...b));
    openSshTab(id);
    const [tab] = sshHostState(id).tabs;

    ssh.__emit(id, tab.sessionId, Uint8Array.from([1, 2]));

    expect(seen).toEqual([1, 2]);
  });

  /**
   * The native side writes the bytes into the terminal itself and reports the
   * event with no payload. Writing again from here would double every
   * character on screen.
   */
  it('does not write fast-path output a second time', () => {
    const id = freshHost();
    const seen: number[] = [];
    attachSshSink(id, b => seen.push(...b));
    openSshTab(id);
    const [tab] = sshHostState(id).tabs;

    ssh.__emit(id, tab.sessionId, new Uint8Array(0), true);

    expect(seen).toEqual([]);
  });

  /**
   * A fast-path event must not queue either. Queuing an empty chunk would
   * hand the next slow-path block to the drain instead of the sink, which is
   * the round trip the fast path exists to avoid.
   */
  it('keeps delivering to the renderer after a fast-path event', () => {
    const id = freshHost();
    const seen: number[] = [];
    attachSshSink(id, b => seen.push(...b));
    openSshTab(id);
    const [tab] = sshHostState(id).tabs;

    ssh.__emit(id, tab.sessionId, new Uint8Array(0), true);
    ssh.__emit(id, tab.sessionId, Uint8Array.from([7]));

    expect(seen).toEqual([7]);
    expect(terminalStore.feed).not.toHaveBeenCalled();
  });

  // Output that arrives while the user is elsewhere has to survive, or coming
  // back shows a terminal missing everything that happened.
  // Output that arrives with no renderer attached goes into the session's own
  // terminal, which owns the scrollback. Nothing is dropped and nothing waits
  // for a view to come back.
  it('writes detached output into the session terminal', async () => {
    const id = freshHost();
    const detach = attachSshSink(id, () => undefined);
    openSshTab(id);
    const [tab] = sshHostState(id).tabs;
    detach();

    ssh.__emit(id, tab.sessionId, Uint8Array.from([7]));
    await flushMicrotasks();

    expect(terminalStore.feed).toHaveBeenCalledWith(
      tab.sessionId,
      expect.any(String),
      expect.any(Number),
      expect.any(Number),
    );
  });

  /**
   * The guarantee this app is built on: output is never discarded, however
   * much of it arrives while the tab is off screen.
   */
  it('keeps every byte of a large detached burst', async () => {
    const id = freshHost();
    const detach = attachSshSink(id, () => undefined);
    openSshTab(id);
    const [tab] = sshHostState(id).tabs;
    detach();

    // Comfortably past the old 256KB cap, which discarded the oldest output.
    const chunk = new Uint8Array(64 * 1024).fill(9);
    for (let i = 0; i < 8; i += 1) ssh.__emit(id, tab.sessionId, chunk);
    await flushMicrotasks();

    const sent = terminalStore.feed.mock.calls
      .filter(c => c[0] === tab.sessionId)
      .reduce((n, c) => n + decodeBase64(c[1] as string).length, 0);
    expect(sent).toBe(8 * chunk.length);
  });

  it('does not send a background tab output to the renderer', () => {
    const id = freshHost();
    const seen: number[] = [];
    attachSshSink(id, b => seen.push(...b));
    openSshTab(id);
    openSshTab(id);
    const [background, active] = sshHostState(id).tabs;
    expect(sshHostState(id).activeSessionId).toBe(active.sessionId);

    ssh.__emit(id, background.sessionId, Uint8Array.from([9]));

    expect(seen).toEqual([]);
  });

  // Selecting must not write through the sink that is still attached: it
  // belongs to the tab being left, and its terminal is about to be replaced.
  // Writing there put the new tab's output into the old tab's screen.
  it('does not write into the outgoing tab on select', () => {
    const id = freshHost();
    const seen: number[] = [];
    attachSshSink(id, b => seen.push(...b));
    openSshTab(id);
    openSshTab(id);
    const [first] = sshHostState(id).tabs;

    ssh.__emit(id, first.sessionId, Uint8Array.from([5]));
    selectSshTab(id, first.sessionId);

    expect(seen).toEqual([]);
  });

  it('writes a background tab output into its own terminal', async () => {
    const id = freshHost();
    openSshTab(id);
    openSshTab(id);
    const [first] = sshHostState(id).tabs;

    ssh.__emit(id, first.sessionId, Uint8Array.from([5]));
    await flushMicrotasks();

    // It went to the hidden tab's terminal, not to whatever is on screen.
    expect(terminalStore.feed).toHaveBeenCalledWith(
      first.sessionId,
      expect.any(String),
      expect.any(Number),
      expect.any(Number),
    );
  });

  /**
   * Order is the whole point of the single queue. Output that arrives while a
   * tab is being switched must not overtake output already queued for it:
   * both end up in the same terminal, and a later chunk landing first paints
   * the newest text and then buries it under the older text.
   */
  it('keeps chunks in order across a tab switch', async () => {
    const id = freshHost();
    openSshTab(id);
    openSshTab(id);
    const [first] = sshHostState(id).tabs;

    // Hold the first write open, so it is still in flight when the tab is
    // switched and the next chunk arrives. Without this the race cannot
    // happen and the test proves nothing.
    let releaseWrite = (): void => {};
    terminalStore.feed.mockImplementationOnce(
      () =>
        new Promise(resolve => {
          releaseWrite = () => resolve(true);
        }),
    );

    ssh.__emit(id, first.sessionId, Uint8Array.from([1]));
    await flushMicrotasks();

    // The tab becomes active and a view attaches while chunk 1 is unfinished.
    selectSshTab(id, first.sessionId);
    const seen: number[] = [];
    attachSshSink(id, b => seen.push(...b));
    ssh.__emit(id, first.sessionId, Uint8Array.from([2]));
    await flushMicrotasks();

    // Chunk 2 must not reach the screen before chunk 1 has been written.
    expect(seen).toEqual([]);

    releaseWrite();
    await flushMicrotasks();
    expect(seen).toEqual([2]);
  });

  /**
   * Numbering by how many tabs are open reuses a number the moment one is
   * closed, so two tabs share a name and the strip looks like it skipped one.
   */
  it('does not reuse a shell number after a tab is closed', () => {
    const id = freshHost();
    openSshTab(id);
    openSshTab(id);
    openSshTab(id);
    const titles = () => sshHostState(id).tabs.map(t => t.title);
    expect(titles()).toEqual(['Shell 1', 'Shell 2', 'Shell 3']);

    const [, second] = sshHostState(id).tabs;
    closeSshTab(id, second.sessionId);
    openSshTab(id);

    const now = titles();
    expect(new Set(now).size).toBe(now.length);
    expect(now).toContain('Shell 2');
  });

  /**
   * A write that did not happen must still reach the tab. Leaving it queued
   * until the next output arrives is what made a tab render late by exactly
   * the amount of output it missed while it was hidden.
   */
  it('delivers a block through the view when its write failed', async () => {
    const id = freshHost();
    openSshTab(id);
    openSshTab(id);
    const [first] = sshHostState(id).tabs;

    // The hidden tab's write fails: nothing reached any terminal.
    let releaseWrite: (ok: boolean) => void = () => {};
    terminalStore.feed.mockImplementationOnce(
      () =>
        new Promise<boolean>(resolve => {
          releaseWrite = resolve;
        }),
    );

    ssh.__emit(id, first.sessionId, Uint8Array.from([1]));
    await flushMicrotasks();

    // The tab becomes active while the write is still unresolved.
    selectSshTab(id, first.sessionId);
    const seen: number[] = [];
    attachSshSink(id, b => seen.push(...b));

    // The write reports failure; the block must go to the view, not wait.
    releaseWrite(false);
    await flushMicrotasks();

    expect(seen).toEqual([1]);
  });

  /** Switching tabs must not replay what the terminal already holds. */
  it('does not resend terminal output when a tab is selected', async () => {
    const id = freshHost();
    openSshTab(id);
    openSshTab(id);
    const [first] = sshHostState(id).tabs;

    ssh.__emit(id, first.sessionId, Uint8Array.from([5]));
    await flushMicrotasks();

    selectSshTab(id, first.sessionId);
    const seen: number[] = [];
    attachSshSink(id, b => seen.push(...b));
    expect(seen).toEqual([]);
  });
});

describe('sizing', () => {
  it('is unsized until the viewport reports a grid', () => {
    const id = freshHost();
    expect(sshHostSized(id)).toBe(false);
    resizeSshHost(id, { cols: 100, rows: 40 });
    expect(sshHostSized(id)).toBe(true);
  });

  it('ignores a zero-sized report', () => {
    const id = freshHost();
    resizeSshHost(id, { cols: 0, rows: 0 });
    expect(sshHostSized(id)).toBe(false);
  });

  // The first measurement only records the size: nothing is connected yet, so
  // sending a resize would be a call against a session that does not exist.
  it('does not resize on the first measurement', () => {
    const id = freshHost();
    resizeSshHost(id, { cols: 100, rows: 40 });
    expect(ssh.resize).not.toHaveBeenCalled();
  });

  // The reason this exists: a session opened against the 80x24 placeholder has
  // a PTY whose row count does not match the screen, and an application that
  // draws with absolute cursor moves puts its overlay in the wrong place.
  it('connects with the measured grid, not the placeholder', () => {
    const id = freshHost();
    resizeSshHost(id, { cols: 100, rows: 40 });
    openSshTab(id);

    expect(ssh.connect).toHaveBeenCalledWith(
      id,
      sshHostState(id).tabs[0].sessionId,
      100,
      40,
    );
  });

  it('resizes every open tab once sized', () => {
    const id = freshHost();
    resizeSshHost(id, { cols: 100, rows: 40 });
    openSshTab(id);
    openSshTab(id);
    (ssh.resize as jest.Mock).mockClear();

    resizeSshHost(id, { cols: 90, rows: 30 });

    expect(ssh.resize).toHaveBeenCalledTimes(2);
  });
});

describe('renaming', () => {
  it('renames the requested tab only', () => {
    const id = freshHost();
    openSshTab(id);
    openSshTab(id);
    const [first, second] = sshHostState(id).tabs;

    renameSshTab(id, first.sessionId, 'build');

    const tabs = sshHostState(id).tabs;
    expect(tabs[0].title).toBe('build');
    expect(tabs[1].title).toBe(second.title);
  });

  it('ignores a blank name', () => {
    const id = freshHost();
    openSshTab(id);
    const before = sshHostState(id).tabs[0].title;

    renameSshTab(id, sshHostState(id).tabs[0].sessionId, '  ');

    expect(sshHostState(id).tabs[0].title).toBe(before);
  });

  it('notifies subscribers', () => {
    const id = freshHost();
    openSshTab(id);
    const listener = jest.fn();
    subscribeSshHost(id, listener);

    renameSshTab(id, sshHostState(id).tabs[0].sessionId, 'renamed');

    expect(listener).toHaveBeenCalled();
  });
});

describe('subscription', () => {
  it('notifies a listener when tabs change', () => {
    const id = freshHost();
    const listener = jest.fn();
    const off = subscribeSshHost(id, listener);

    openSshTab(id);
    expect(listener).toHaveBeenCalled();

    off();
    listener.mockClear();
    openSshTab(id);
    expect(listener).not.toHaveBeenCalled();
  });
});

File: app/android/app/src/test/java/com/remotly/app/session/SshSessionsTest.kt
package com.remotly.app.session

import java.util.concurrent.atomic.AtomicInteger
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test

class SshSessionsTest {

    private lateinit var ssh: FakeSshTransport

    @Before
    fun setUp() {
        ssh = FakeSshTransport()
        SshSessions.transport = ssh
    }

    // Each test gets its own host id: the store is a process-wide singleton
    // that outlives every test, and JUnit4 makes a fresh instance of this
    // class per method, so the counter has to live in the companion to stay
    // unique across the whole run rather than resetting every method.
    private fun freshHost(): String = "h${hostSeq.incrementAndGet()}"

    companion object {
        private val hostSeq = AtomicInteger()
    }

    private fun tabsOf(hostId: String) = SshSessions.state(hostId).value.tabs

    // -- session lifetime --------------------------------------------------------

    @Test
    fun `does not reopen a tab for a host that already has one`() {
        val id = freshHost()
        SshSessions.openTab(id)
        assertTrue(SshSessions.hostStarted(id))
        assertEquals(1, tabsOf(id).tabs.size)
    }

    @Test
    fun `closes exactly the tab that was asked for`() {
        val id = freshHost()
        SshSessions.openTab(id)
        SshSessions.openTab(id)
        val (first, second) = tabsOf(id).tabs

        SshSessions.closeTab(id, first.sessionId)

        assertEquals(id to first.sessionId, ssh.closeCalls.single())
        assertEquals(listOf(second.sessionId), tabsOf(id).tabs.map { it.sessionId })
    }

    @Test
    fun `closes every tab when the host disconnects`() {
        val id = freshHost()
        SshSessions.openTab(id)
        SshSessions.openTab(id)

        SshSessions.closeHost(id)

        assertEquals(2, ssh.closeCalls.size)
        assertTrue(tabsOf(id).tabs.isEmpty())
    }

    // -- sizing --------------------------------------------------------

    @Test
    fun `is unsized until the viewport reports a grid`() {
        val id = freshHost()
        assertFalse(SshSessions.hostSized(id))
        SshSessions.resizeHost(id, 100, 40)
        assertTrue(SshSessions.hostSized(id))
    }

    @Test
    fun `ignores a zero-sized report`() {
        val id = freshHost()
        SshSessions.resizeHost(id, 0, 0)
        assertFalse(SshSessions.hostSized(id))
    }

    @Test
    fun `does not resize before any tab is open`() {
        val id = freshHost()
        SshSessions.resizeHost(id, 100, 40)
        assertTrue(ssh.resizeCalls.isEmpty())
    }

    // The reason this matters: a session opened against a placeholder grid
    // gets a pty whose row count does not match the screen, and a program
    // that draws with absolute cursor moves puts its overlay in the wrong
    // place.
    @Test
    fun `connects with the measured grid, not the placeholder`() {
        val id = freshHost()
        SshSessions.resizeHost(id, 100, 40)
        SshSessions.openTab(id)

        val call = ssh.connectCalls.single()
        assertEquals(tabsOf(id).tabs[0].sessionId, call.sessionId)
        assertEquals(100, call.cols)
        assertEquals(40, call.rows)
    }

    @Test
    fun `resizes every open tab once sized`() {
        val id = freshHost()
        SshSessions.resizeHost(id, 100, 40)
        SshSessions.openTab(id)
        SshSessions.openTab(id)
        ssh.resizeCalls.clear()

        SshSessions.resizeHost(id, 90, 30)

        assertEquals(2, ssh.resizeCalls.size)
    }

    // -- renaming --------------------------------------------------------

    @Test
    fun `renames the requested tab only`() {
        val id = freshHost()
        SshSessions.openTab(id)
        SshSessions.openTab(id)
        val (first, second) = tabsOf(id).tabs

        SshSessions.renameTab(id, first.sessionId, "build")

        val tabs = tabsOf(id).tabs
        assertEquals("build", tabs[0].title)
        assertEquals(second.title, tabs[1].title)
    }

    @Test
    fun `ignores a blank rename`() {
        val id = freshHost()
        SshSessions.openTab(id)
        val before = tabsOf(id).tabs[0].title

        SshSessions.renameTab(id, tabsOf(id).tabs[0].sessionId, "  ")

        assertEquals(before, tabsOf(id).tabs[0].title)
    }

    // -- subscription --------------------------------------------------------

    @Test
    fun `publishes a new state value when a tab opens`() {
        val id = freshHost()
        val flow = SshSessions.state(id)
        val before = flow.value

        SshSessions.openTab(id)

        assertTrue(flow.value !== before)
        assertEquals(1, flow.value.tabs.tabs.size)
    }

    @Test
    fun `keeps one host's state independent of another's`() {
        val a = freshHost()
        val b = freshHost()
        SshSessions.openTab(a)
        assertTrue(tabsOf(b).tabs.isEmpty())
    }

    // -- host key prompts --------------------------------------------------------

    @Test
    fun `surfaces a host key prompt from the transport`() {
        val id = freshHost()
        SshSessions.openTab(id)
        val sessionId = tabsOf(id).tabs[0].sessionId

        ssh.emitState(id, sessionId, mapOf("state" to "hostKey", "algorithm" to "ed25519", "fingerprint" to "SHA256:abc", "changed" to false))

        val state = SshSessions.state(id).value
        assertEquals(SshTabPhase.HostKey, state.tabs.tabs[0].phase)
        assertEquals(sessionId, state.hostKeyPrompt?.sessionId)
        assertEquals("SHA256:abc", state.hostKeyPrompt?.fingerprint)
    }

    @Test
    fun `accepting a host key clears the prompt and forwards the decision`() {
        val id = freshHost()
        SshSessions.openTab(id)
        val sessionId = tabsOf(id).tabs[0].sessionId
        ssh.emitState(id, sessionId, mapOf("state" to "hostKey", "algorithm" to "", "fingerprint" to "", "changed" to false))

        SshSessions.answerHostKey(id, HostKeyDecision.Accept)

        assertNull(SshSessions.state(id).value.hostKeyPrompt)
        assertEquals(Triple(id, sessionId, "accept"), ssh.hostKeyCalls.single())
        assertTrue(ssh.closeCalls.isEmpty())
    }

    @Test
    fun `rejecting a host key closes the session instead of forwarding it`() {
        val id = freshHost()
        SshSessions.openTab(id)
        val sessionId = tabsOf(id).tabs[0].sessionId
        ssh.emitState(id, sessionId, mapOf("state" to "hostKey", "algorithm" to "", "fingerprint" to "", "changed" to false))

        SshSessions.answerHostKey(id, HostKeyDecision.Reject)

        assertNull(SshSessions.state(id).value.hostKeyPrompt)
        assertTrue(ssh.hostKeyCalls.isEmpty())
        assertEquals(id to sessionId, ssh.closeCalls.single())
        val tab = tabsOf(id).tabs[0]
        assertEquals(SshTabPhase.Closed, tab.phase)
        assertEquals("The host key was rejected.", tab.detail)
    }

    // -- a tab opened with a command --------------------------------------------------------

    @Test
    fun `types the queued command once the shell reports ready, and not before`() {
        val id = freshHost()
        SshSessions.openTab(id, title = "api", runs = "herdr")
        val sessionId = tabsOf(id).tabs[0].sessionId
        assertTrue(ssh.writeCalls.isEmpty())

        ssh.emitState(id, sessionId, mapOf("state" to "active"))

        val write = ssh.writeCalls.single()
        assertEquals("herdr\n", String(write.bytes, Charsets.UTF_8))
    }

    @Test
    fun `does not type the command again on a later ready`() {
        val id = freshHost()
        SshSessions.openTab(id, title = "api", runs = "herdr")
        val sessionId = tabsOf(id).tabs[0].sessionId

        ssh.emitState(id, sessionId, mapOf("state" to "active"))
        ssh.emitState(id, sessionId, mapOf("state" to "active"))

        assertEquals(1, ssh.writeCalls.size)
    }

    @Test
    fun `keeps the given name when the program repaints the title`() {
        val id = freshHost()
        SshSessions.openTab(id, title = "api", runs = "herdr")

        SshSessions.reportTerminalTitle(id, "host: api")

        assertEquals("api", tabsOf(id).tabs[0].title)
    }

    @Test
    fun `reveals the terminal a workspace session is already attached in`() {
        val id = freshHost()
        SshSessions.openWorkspaceTab(id, workspaceId = "w1", label = "api", runs = "herdr")
        val first = tabsOf(id).tabs[0].sessionId
        ssh.emitState(id, first, mapOf("state" to "active"))
        SshSessions.openTab(id)
        SshSessions.selectTab(id, tabsOf(id).tabs[1].sessionId)

        SshSessions.openWorkspaceTab(id, workspaceId = "w1", label = "api", runs = "herdr")

        assertEquals(2, tabsOf(id).tabs.size)
        assertEquals(first, tabsOf(id).activeSessionId)
    }

    @Test
    fun `retags the terminal when another workspace is entered`() {
        val id = freshHost()
        SshSessions.openWorkspaceTab(id, workspaceId = "w1", label = "api", runs = "herdr")
        val first = tabsOf(id).tabs[0].sessionId
        ssh.emitState(id, first, mapOf("state" to "active"))

        SshSessions.openWorkspaceTab(id, workspaceId = "w2", label = "deploy", runs = "herdr")

        assertEquals(1, tabsOf(id).tabs.size)
        assertEquals("w2", tabsOf(id).tabs[0].workspaceId)
        assertEquals("deploy", tabsOf(id).tabs[0].title)
    }

    @Test
    fun `leaves the shell budget alone for a workspace tab`() {
        val id = freshHost()
        SshSessions.openWorkspaceTab(id, workspaceId = "w1", label = "api", runs = "herdr")
        ssh.emitState(id, tabsOf(id).tabs[0].sessionId, mapOf("state" to "active"))

        repeat(MAX_SSH_TABS) { SshSessions.openTab(id) }

        assertEquals(MAX_SSH_TABS, tabsOf(id).tabs.count { it.kind == SshTabKind.Shell })
        assertFalse(SshSessions.canAddTab(id))
    }

    @Test
    fun `opens a workspace tab over a full shell strip`() {
        val id = freshHost()
        repeat(MAX_SSH_TABS) { SshSessions.openTab(id) }

        val opened = SshSessions.openWorkspaceTab(id, workspaceId = "w1", label = "api", runs = "herdr")

        assertNotNull(opened)
        assertEquals(1, tabsOf(id).tabs.count { it.kind == SshTabKind.Workspace })
    }

    @Test
    fun `opens a new workspace terminal once the old one has closed`() {
        val id = freshHost()
        SshSessions.openWorkspaceTab(id, workspaceId = "w1", label = "api", runs = "herdr")
        ssh.emitState(id, tabsOf(id).tabs[0].sessionId, mapOf("state" to "closed", "userInitiated" to false))

        SshSessions.openWorkspaceTab(id, workspaceId = "w1", label = "api", runs = "herdr")

        assertEquals(2, tabsOf(id).tabs.size)
    }

    @Test
    fun `gives a second herdr session its own terminal`() {
        val id = freshHost()
        SshSessions.openWorkspaceTab(id, workspaceId = "w1", label = "api", runs = "herdr")
        ssh.emitState(id, tabsOf(id).tabs[0].sessionId, mapOf("state" to "active"))

        SshSessions.openWorkspaceTab(id, workspaceId = "w1", label = "api", runs = "herdr --session work", session = "work")

        assertEquals(2, tabsOf(id).tabs.size)
    }

    // -- bare session id binding --------------------------------------------------------
    //
    // Two documented bugs came from a screen or a release call keying a
    // terminal by `hostId:sessionId` instead of the bare id: a view adopted
    // an empty terminal while the tab's real screen sat in one nobody drew,
    // and a closed tab's terminal was never released because nothing was
    // retained under the composite key. Every call this store makes into the
    // transport must carry the same bare id the tab itself carries.

    @Test
    fun `connects using the tab's own bare session id`() {
        val id = freshHost()
        SshSessions.openTab(id)
        val sessionId = tabsOf(id).tabs[0].sessionId

        assertFalse(sessionId.contains(":"))
        assertEquals(sessionId, ssh.connectCalls.single().sessionId)
    }

    @Test
    fun `closes using the tab's own bare session id, never a host-qualified one`() {
        val id = freshHost()
        SshSessions.openTab(id)
        val sessionId = tabsOf(id).tabs[0].sessionId

        SshSessions.closeTab(id, sessionId)

        val (closedHost, closedSession) = ssh.closeCalls.single()
        assertEquals(id, closedHost)
        assertEquals(sessionId, closedSession)
        assertFalse(closedSession.contains(":"))
        assertFalse(closedSession.contains(id))
    }

    @Test
    fun `a files tab opens no connection and is usable at once`() {
        val id = freshHost()
        SshSessions.openTab(id, kind = SshTabKind.Files)

        assertTrue(ssh.connectCalls.isEmpty())
        assertEquals(SshTabPhase.Active, tabsOf(id).tabs.single().phase)
        assertEquals("Files", tabsOf(id).tabs.single().title)
    }

    @Test
    fun `closing a files tab never closes a channel it never opened`() {
        val id = freshHost()
        SshSessions.openTab(id)
        val shell = tabsOf(id).tabs.single().sessionId
        SshSessions.openTab(id, kind = SshTabKind.Files)
        val files = tabsOf(id).tabs.first { it.sessionId != shell }.sessionId

        SshSessions.closeTab(id, files)

        assertTrue(ssh.closeCalls.isEmpty())
        assertEquals(shell, tabsOf(id).tabs.single().sessionId)
    }

    @Test
    fun `writes input under the active tab's bare session id`() {
        val id = freshHost()
        SshSessions.openTab(id)
        val sessionId = tabsOf(id).tabs[0].sessionId

        SshSessions.sendInput(id, byteArrayOf(1, 2, 3))

        assertEquals(sessionId, ssh.writeCalls.single().sessionId)
    }
}
